src/services/QuoteService.ts

```diff
@@ -1,12 +1,241 @@
+} from '../repositories/ModifierRepository';
+
+export interface QuoteRepositories {
+  ratesRepository: SpaceRateRepository;
+  spaceModifiersRepository: SpaceModifierRepository;
+  facilityModifierRepository: FacilityModifierRepository;
+}
 
 export class QuoteService {
+  // Returns a quote for the Ask
   public quote = async (
     facilityId: string,
     spaceId: string,
     ask: Ask
   ): Promise<BigNumber> => {
-    throw new Error('Not Implemented Yet');
+    if (!ask.checkIn) {
+      throw new Error('Invalid checkIn date');
+    }
+
+    if (!ask.checkOut) {
+      throw new Error('Invalid checkOut date');
+    }
+
+    const repositories: QuoteRepositories = {
+      ratesRepository: new SpaceRateRepository(facilityId, spaceId),
+      spaceModifiersRepository: new SpaceModifierRepository(
+        facilityId,
+        spaceId
+      ),
+      facilityModifierRepository: new FacilityModifierRepository(facilityId)
+    };
+    let total = BigNumber.from(0);
+
+    // Iterate through days from ask.checkIn to ask.checkOut
+    for await (const value of this.getQuoteIterator(
+      repositories,
+      ask.checkIn,
+      ask.checkOut
+    )) {
+      total = total.add(value);
+    }
+
+    // Apply length_of_stay modifier, in order of priority
+    let losSpaceModifier: undefined | LOSRateModifier;
+
+    try {
+      losSpaceModifier =
+        (await repositories.spaceModifiersRepository.getModifier(
+          'length_of_stay'
+        )) as LOSRateModifier;
+    } catch (e) {
+      if (e.status !== 404) {
+        throw e;
+      }
+
+      try {
+        losSpaceModifier =
+          (await repositories.facilityModifierRepository.getModifier(
+            'length_of_stay'
+          )) as LOSRateModifier;
+      } catch (e) {
+        if (e.status !== 404) {
+          throw e;
+        }
+      }
+    }
+
+    if (losSpaceModifier) {
+      const daysTotal = DateTime.fromObject(ask.checkOut).diff(
+        DateTime.fromObject(ask.checkIn)
+      ).days;
+      let applicable = false;
+
+      switch (losSpaceModifier.condition) {
+        case 0:
+          applicable = daysTotal < losSpaceModifier.los;
+          break;
+        case 1:
+          applicable = daysTotal <= losSpaceModifier.los;
+          break;
+        case 2:
+          applicable = daysTotal === losSpaceModifier.los;
+          break;
+        case 3:
+          applicable = daysTotal >= losSpaceModifier.los;
+          break;
+        case 4:
+          applicable = daysTotal > losSpaceModifier.los;
+          break;
+        default:
+      }
+
+      if (applicable) {
+        switch (losSpaceModifier.valueOneof.oneofKind) {
+          case 'ratio':
+            total = total
+              .mul(BigNumber.from(losSpaceModifier.valueOneof.ratio.p))
+              .div(BigNumber.from(losSpaceModifier.valueOneof.ratio.q));
+            break;
+          case 'fixed':
+            total = BigNumber.from(losSpaceModifier.valueOneof.fixed);
```

Comment:
`fixed` is to be interpreted as a "fixed" _modifier_, so this represents a constant either addition or subtraction. The value specified in the protobuf is an `int`, so by definition may be positive or negative. Therefore:

`total = total.add(BigNumber.from(losSpaceModifier.valueOneof.fixed));`

Reply:
fiixed